oauth2client/flask_util.py

```diff
@@ -0,0 +1,403 @@
+        bp.add_url_rule('/oauth2logout', 'logout', self.logout_view)
+        bp.add_url_rule('/oauth2callback', 'callback', self.callback_view)
+
+        return bp
+
+    def authorize_view(self):
+        args = request.args.to_dict()
+
+        return_url = args.pop('return_url', None)
+        if return_url is None:
+            return_url = request.referrer or '/'
+        session['google_oauth2_return'] = return_url
+
+        flow = self._make_flow(**args)
+        auth_url = flow.step1_get_authorize_url()
+
+        return redirect(auth_url)
+
+    def logout_view(self):
+        self.storage.delete()
+
+        if self.logout_callback:
+            self.logout_callback()
+
+        return redirect(request.referrer or '/')
+
+    def callback_view(self):
+        if 'error' in request.args:
+            reason = request.args.get(
+                'error_description', request.args.get('error', ''))
+            return 'Authorization failed: %s' % reason, httplib.BAD_REQUEST
+
+        try:
+            client_csrf = request.args['state']
+            server_csrf = session['google_oauth2_state']
+            code = request.args['code']
+        except KeyError:
+            return 'Invalid request', httplib.BAD_REQUEST
+
+        if client_csrf != server_csrf:
+            return 'Invalid request state', httplib.BAD_REQUEST
+
+        flow = self._make_flow()
+
+        # Exchange the auth code for credentials.
+        try:
+            credentials = flow.step2_exchange(code)
+        except FlowExchangeError as exchange_error:
+            current_app.logger.exception(exchange_error)
+            return 'An error occurred: %s' % exchange_error, httplib.BAD_REQUEST
+
+        # Save the credentials to the storage.
+        self.storage.put(credentials)
+
+        if self.authorize_callback:
+            self.authorize_callback(credentials)
+
+        return redirect(session.pop('google_oauth2_return', '/'))
+
+    @property
+    def credentials(self):
+        ctx = _app_ctx_stack.top
+
+        if not hasattr(ctx, 'google_oauth2_credentials'):
+            ctx.google_oauth2_credentials = self.storage.get()
+
+        return ctx.google_oauth2_credentials
+
+    def has_credentials(self):
+        return self.credentials and not self.credentials.invalid
+
+    @property
+    def email(self):
+        """Returns the user's email address.
+
+        The email address is provided by the current credentials' id_token. The
+        use can change their email, it's more appropriate to use user_id for
+        a unique, user-specific identifier.
+        """
+        if not self.credentials:
+            return None
+        try:
+            return self.credentials.id_token['email']
+        except KeyError:
+            current_app.logger.error(
+                'Invalid id_token %s', self.credentials.id_token)
+
+    @property
+    def user_id(self):
+        """Returns the user's email address.
```

Comment:
Are you quite certain that the property user_id "returns the user's email address"?